core: add context parameter to k8sutil node

y1r opened the pull request:
**Description of your changes:**

This commit adds context parameter to k8sutil node functions. By this,
we can handle cancellation during API call of node resource.

Signed-off-by: Yuichiro Ueno <y1r.ueno@gmail.com>

**Which issue is resolved by this Pull Request:**
Part of https://github.com/rook/rook/issues/8700

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.
- [x] Pending release notes updated with breaking and/or notable changes, if necessary.
- [x] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [x] Code generation (`make codegen`) has been run to update object specifications, if necessary.

leseb commented on pkg/operator/ceph/cluster/osd/osd.go:
```diff
@@ -601,7 +601,7 @@ func getLocationFromPod(clientset kubernetes.Interface, d *appsv1.Deployment, cr
 	}
 	nodeName := pods.Items[0].Spec.NodeName
-	hostName, err := k8sutil.GetNodeHostName(clientset, nodeName)
+	hostName, err := k8sutil.GetNodeHostName(ctx, clientset, nodeName)
```
This is still using `ctx := context.TODO()`, see L598.

leseb commented on pkg/operator/discover/discover.go:
```diff
@@ -255,7 +255,7 @@ func ListDevices(clusterdContext *clusterd.Context, namespace, nodeName string)
 	if len(nodeName) > 0 {
 		var err error
-		nodeName, err = k8sutil.GetNodeNameFromHostname(clusterdContext.Clientset, nodeName)
+		nodeName, err = k8sutil.GetNodeNameFromHostname(ctx, clusterdContext.Clientset, nodeName)
```
This is still using `ctx := context.TODO()`.

y1r commented:
@leseb I've updated the PR, could you review it again?

y1r commented:
@leseb I've updated the PR, so could you review this again?
- checked `pkg/operator/discover/discover.go` and removed all context.TODO
- checked `pkg/operator/ceph/cluster/osd/osd.go` and removed all context.TODO

leseb commented on cmd/rook/ceph/osd.go:
```diff
@@ -214,9 +214,11 @@ func prepareOSD(cmd *cobra.Command, args []string) error {
 	context := createContext()
 	commonOSDInit(provisionCmd)
-	crushLocation, topologyAffinity, err := getLocation(context.Clientset)
+	crushLocation, topologyAffinity, err := getLocation(backgroundCtx, context.Clientset)
```
Why not use `cmd.Context()` like done in the other PR?

y1r replied:
I've updated the PR, I updated `prepareOSD` and `removeOSDs`.

leseb reviewed (commented):
Small question

leseb commented:
@y1r this is the last PR in the series to complete https://github.com/rook/rook/issues/8700 right? So I edited the description of the PR and swapped `part of` with `closes`. Thanks

y1r commented:
> this is the last PR in the series to complete #8700 right?

Not yet. I executed grep on the top of rebased this PR:
```
$ grep -R 'context.TODO()' . | grep -v test.go | grep -v tests | grep context.TODO | sort
./configmap.go: ctx := context.TODO()
./endpoint.go:  ctx := context.TODO()
./kvstore.go:   ctx := context.TODO()
./kvstore.go:   ctx := context.TODO()
./kvstore.go:   ctx := context.TODO()
./kvstore.go:   ctx := context.TODO()
./prometheus.go:        ctx := context.TODO()
./prometheus.go:        ctx := context.TODO()
./pvc.go:               err := client.Get(context.TODO(), types.NamespacedName{Name: *(currentPVC.Spec.StorageClassName)}, storageClass)
./pvc.go:               if err = client.Update(context.TODO(), currentPVC); err != nil {
./replicaset.go:        ctx := context.TODO()
./secret.go:    ctx := context.TODO()
./service.go:   ctx := context.TODO()
./service.go:   ctx := context.TODO()
./service.go:   ctx := context.TODO()
```

I'll do them later.

leseb commented:
> > this is the last PR in the series to complete #8700 right?
> 
> Not yet. I executed grep on the top of rebased this PR:
> 
> ```
> $ grep -R 'context.TODO()' . | grep -v test.go | grep -v tests | grep context.TODO | sort
> ./configmap.go: ctx := context.TODO()
> ./endpoint.go:  ctx := context.TODO()
> ./kvstore.go:   ctx := context.TODO()
> ./kvstore.go:   ctx := context.TODO()
> ./kvstore.go:   ctx := context.TODO()
> ./kvstore.go:   ctx := context.TODO()
> ./prometheus.go:        ctx := context.TODO()
> ./prometheus.go:        ctx := context.TODO()
> ./pvc.go:               err := client.Get(context.TODO(), types.NamespacedName{Name: *(currentPVC.Spec.StorageClassName)}, storageClass)
> ./pvc.go:               if err = client.Update(context.TODO(), currentPVC); err != nil {
> ./replicaset.go:        ctx := context.TODO()
> ./secret.go:    ctx := context.TODO()
> ./service.go:   ctx := context.TODO()
> ./service.go:   ctx := context.TODO()
> ./service.go:   ctx := context.TODO()
> ```
> 
> I'll do them later.

Ok I reverted my edit, thanks!
